README.md

```diff
@@ -1,16 +1,1602 @@
+
+ send.result.ok: success
+ send.result.error: { error }
+}
+
+
+send.run --> send.task
+send.auth --> auth
+
+read.run --> read.task
+read.auth --> auth
+
+update.run --> update.task
+update.auth --> auth
+
+scope --> send.task
+scope --> read.task
+scope --> update.task
+
+
+send.promise.await --> send: ok
+send.promise.await --> send: error
+send.promise.await --> send: *
+
+read.promise.await --> read.task: ok
+read.promise.await --> read.task: error
+read.promise.await --> read.task: *
+
+
+send.receipt.out --> send.result
+send.receipt.sig --> send.receipt.ran
+send.receipt.sig --> send.receipt.out
+send.receipt.sig --> send.receipt.fx
+send.receipt.sig --> send.receipt.meta
+send.receipt.sig --> send.receipt.iss
+send.receipt.sig --> send.receipt.prf
+
+
+send.receipt.ran --> send
+```
+
+### 2.2.1 Task
+
+A [Task] is like a deferred function application: a request to perform some action on a resource with specific input.
+
+### 2.2.2 Authorization
+
+An [Authorization] is a cryptographically signed proof permitting execution of referenced tasks. It allows invoker to authorize a group of tasks using one cryptographic signature.
+
+### 2.2.3 Invocation
+
+An [Invocation] is an invoker authorized instruction to an executor to run the [Task].
+
+### 2.2.4 Result
+
+A [Result] is the output of a [Task].
+
+### 2.2.5 Receipt
+
+A [Receipt] is a cryptographically signed description of the [Invocation] output.
+
+### 2.2.6 Promise
+
+A [promise] is a reference to an eventual [Receipt] of an [Invocation].
+
+## 2.3 IPLD Schema
+
+```ipldsch
+type Task<Args> struct {
+  v       SemVer
+  with    URI
+  do      Ability
+  input   Args (implicit {})
+  meta    {String : Any} (implicit {})
+  nnc     string (implicit "")
+  prf     [&UCAN] (implicit [])
+}
+
+type SemVer string
+type URI string
+
+type Authorization struct {
+  # Authorization is denoted by the set of links been authorized
+  scope   [&Any] (implicit [])
+  # Scope signed by the invoker
+  s       VarSig
+}
+
+type Invocation<In> struct {
+  task &Task<In>
+  authorization  &Authorization
+} representation tuple
+
+type Receipt<In, Ok, Error> struct {
+  ran     &Invocation<In>
+
+  # output of the invocation
+  out     Result<Ok, Error>
+  # Effects to be performed
+  fx      [&Invocation<Any>] (implicit {})
+
+  # Related receipts
+  origin  optional &Receipt<In, Any, Any>
+
+  # All the other metadata
+  meta    { String: Any } (implicit {})
+
+  # Principal that issued this receipt. If omitted issuer is
+  # inferred from the invocation task audience.
+  iss     optional Principal
+
+  # When issuer is different from executor this MUST hold a UCAN
+  # delegation chain from executor to the issuer. Should be omitted executor is an issuer.
+  prf     [&UCAN] implicit ([])
+
+  # Signature from the `iss`.
+  s       Varsig
+}
+
+type Result<Ok, Error> union {
+  | Ok ("ok") # Success
+  | Error ("error") # Error
+} representation kinded
+
+type Promise<In> union {
```

Comment:
After pondering more on the effects, fork/join etc after our conversation I'm think following feels right to me:

```ipldsch
type Select {
  # Join a single invocation
  &Promise          "join"
  # Fork a single invocation
  &Promise          "fork"
} representation keyed

type Effect struct {
  # Fork a single invocation
  &Promise          "fork"
  # Join a single invocation
  &Promise          "join"

  # Group of Fork / Join invocations
  {String: Select}  "select"
} representation keyed

type Receipt struct {
  fx Effect
```

Here is the rational of the above:

1. Simple continuation is just a `{join: ...}` effect. This would be most of web3.storage cases.
2. Case of the single detached task can be expressed using `{fork: ...}`, it's only semantically different from `{join: ...}`. While just `join` tells receipt recipient that effect is a  continuation of the workflow while `fork` tells recipient that it is not.

   Obviously recipient can choose to follow the effect regardless if it's tagged `fork` or a `join`. In web3.storage case however when you'll subscribe to the workflow it will report state updates on "joined" effects, but won't on the "forked" effects.

3. Select is perhaps the interesting effect branch. It is basically allows recipient to choose which effects to follow. It is similar to original invocation spec and allows you to name sub tasks as you're pleased. Idea is that executor is instructed to run all the effects concurrently while tagging which ones meant to be `fork`s and which one meant to be `join`s. 

Thinking here is that it can tell recipient all the effects caused by the task. I think here IPVM will let recipient to choose which task to joint and which ones to not care about. In case of web3.storage `fork` / `join` tags will tell which tasks are considered part of continuation and which tasks are not. It also would allow us to implement system capability e.g `{ do: 'effect/select', with: 'ipld://' + effect.cid }` from which we'd be able to derive joined result of all `join` tagged tasks.

Reply:
@expede I'm going to go back and still say that I think it would make a lot more sense to me if we instead we had an `Task` subtype `Fork` e.g.

```ipldsch
type Fork {
  do "./fork"
  with URI
  input  {String: &Promise}
}
```

I don't think a wrapper with couple fields namely `do, with, v` is significant overhead over just an array. But on the upside this wrapper would give us:

1. `Task` and consequently an `Invocation` by which one could lookup a `Receipt` if the would want to see combined result off all the tasks. If you want to pick and choose it's straight forward as well just pick ones you care from the `input` of the Fork.
2. It would also mean that we'll define `Result` type of the `Fork` task invocation. E.g. We could say that `Fork` result is `ok` as long as all the non `_` prefixed tasks pass and value is map of Results, if some non `_` prefixed tasks fail then result is `error` branch with the same value as in `ok` branch.  


On the flipside it would imply that one could fork tasks as input of the task returned in an effect, which does mean a bit more traversal for the executor, but on the other hand input can already take bunch of tasks as inputs so I'm not sure this really would add any complexity in practice.

Doing opposite and having `Join` as `Task` subtype unfortunately fails to deliver upsites of the `Fork` approach namely:

1. There is no obvious way to ask for the next `Receipt` when several effects are returned. So choices there boil down to pick and choose and then derive some `Join` task to lookup `Receipt` by which you still could do with a `Fork` option.
2. Specifying `Result` type does not apply here because there is no `Task` in this case.
3. If there are more than one `Join` effect you're no longer able to ask for combined result without building up custom `Join` task of all the `Join`s.

Reply:
I end up iterating more over this here https://github.com/ucan-wg/invocation/pull/10#issuecomment-1411317360

Reply:
> There is no obvious way to ask for the next Receipt when several effects are returned. 

I think this may be the biggest difference for the two approaches. IPVM _never_ wants to do this, but I think your use case involves treating each Invocation as steps in a state machine.

IPVM runs an Invocation to completion, and enqueuing the a next Invocation is not even guaranteed to run since it needs to negotiate for resources and trust.

> There is no obvious way to ask for the next Receipt when several effects are returned

Isn't this only the case if where there's an implicit `main` thread that's always being returned to? There's a lot of call graphs that don't return to a single thread of control, especially in a decentralized system with more than many organizations and autonomous agents involved.

I think in a lot of ways you're looking at things as a "push forward" / imperative semantics, and I'm looking for a demand-driven or declarative semantics, which changes how we look at the problem.

Perhaps this difference is part of why I'm struggling to understand the reasoning for having `origin` in the Receipt as opposed to in the `Invocation`?

> Task and consequently an Invocation by which one could lookup a Receipt if the would want to see combined result off all the tasks [...] Doing opposite and having Join as Task subtype unfortunately fails to deliver upsites of the Fork

Right, my challenge with it is that I'm going to have a lot of cases where the effects do not depend on each other, should fail separately, and not be tracked together. If we get a Receipt for all of them together, I will immedietly throw out the Receipt. I have no reason to put it on the DHT in the same way that I don't have no reason to put all of a day's invocations from the same user in a container to track together.

I can have a task type that's "disjoint tasks, treat these separately" — but I'm unclear what the advantage of having that is over an array other than adding nesting?

It feels very strange to have a structure that has no semantics.

> If there are more than one Join effect you're no longer able to ask for combined result without building up custom Join task of all the Joins.

Right, a lot like what IPVM will have to do with workflows/transactions, which AFAICT is essentially a `Join` task type. I wonder if there's a deeper thing in here that would serve both of our use cases?

Reply:
Maybe for my own understanding @gozala. Here's two diagrams:

Together:

```mermaid
graph TB
  subgraph Alice
  A -. emit .-> F
  A -. emit .-> G
  B -- await --> A
  B -- await --> G
  C -- await --> A
  end
```

Split:

```mermaid
graph TB
  subgraph Alice
  A -. emit .-> F
  A -. emit .-> G
  end

  subgraph Bob
  B -- await --> A
  B -- await --> G
  end

  subgraph Carol
  C -- await --> A
  end
```

Absent some kind of transaction wrapper, do you expect these to have the same or different semantics? If different, does the time of invocation enter into that equation?

Reply:
> There is no obvious way to ask for the next Receipt when several effects are returned. So choices there boil down to pick and choose and then derive some Join task to lookup Receipt by which you still could do with a Fork option.

Are you maybe trying to collect all of the tasks emitted in various Invocations into one  "session"? If so, Ceramic does something similar with the blockchain.

Much like a `transaction` or `fork` task type, why not have a `tracking` task type? This would be more flexible than depending on a workflow being a straight-line (it could involve multiple users, be suspended and resumed over long periods, moved between services, etc).